src/services/summarization/summarization.service.ts

```diff
@@ -0,0 +1,59 @@
+import { OnDestroy, Injectable } from '@angular/core';
+import { takeUntil, shareReplay } from 'rxjs/operators';
+import { Subject, Observable } from 'rxjs';
+import { Point } from './summarization-data-source.service';
+import { SummaryGroup } from './types';
+
+export interface BaseConfig {
+  datumLabels: string[];
+}
+
+@Injectable({
+  providedIn: 'any',
+})
+export abstract class SummarizationService<PointT extends Point, PropertiesT, ConfigT extends BaseConfig> implements OnDestroy {
+  destroy$ = new Subject();
+  cachedDataProperties$ = new Map<string, Observable<PropertiesT>>();
+  cachedSummaries$ = new Map<string, Observable<SummaryGroup[]>>();
+
+  hashObject(obj: Record<string, any>): string {
+    return JSON.stringify(obj, Object.keys(obj).sort());

```

Comment:
curious, why do you need to filter and sort it?

Reply:
so that `config = { a, b }` and `config = { b, a }` will return the same hashkey